### PR TITLE
WT-5451 Update reverse modify logic to always insert a modify into the history store

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: src/history/hs.c [hunk 1]
```diff
@@ -669,39 +636,25 @@ __wt_hs_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_RECONCILE *r, WT_M
...
         full_value->data = upd->data;
         full_value->size = upd->size;
 
         squashed = false;
 
-        /* Flush the updates on stack. */
-        for (; modifies.size > 0; tmp = full_value, full_value = prev_full_value,
-                                  prev_full_value = tmp, upd = prev_upd) {
+        /* Flush the updates on stack. Stopping once we run out or we reach the onpage upd txnid. */
+        for (; modifies.size > 0 && upd->txnid != list->onpage_upd->txnid;
```
Comment (quchenhao) [inline src/history/hs.c]: Why upd->txnid != list->onpage_upd->txnid is necessary? upd->txnid can only equal to list->onpage_upd->txnid when modifies is empty, am I correct?

Comment (luke-pearson) [reply src/history/hs.c]: No modifies could contain multiple modifications belonging to the onpage update txnid. i.e. when we insert 3 modifies in the one txn.

Comment (quchenhao) [reply src/history/hs.c]: Oh, I see this is to squash the onpage txn. Can you add a comment for this?

File: src/history/hs.c [hunk 2]
```diff
@@ -766,24 +720,14 @@ __wt_hs_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_RECONCILE *r, WT_M
                 squashed = true;
         }
 
+        if (modifies.size > 0)
+            WT_STAT_CONN_INCR(session, cache_hs_write_squash);
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/history/hs.c]: Now this makes sense. Thank you.

File: src/history/hs.c (hunk 1 above)
Comment (luke-pearson) [reply src/history/hs.c]: Sure

Comment (quchenhao) [reply src/history/hs.c]: I think we need to compare both txnid and timestamp.

**Comment by luke-pearson** [comment]
@quchenhao I had to redisable the test_hs01 as it was broken by merging in pm-1521. I've created a follow up jira to fix this: WT-5510

**Comment by luke-pearson** [comment]
evergreen retry

File: src/history/hs.c [hunk 3]
```diff
@@ -561,39 +531,30 @@ __wt_hs_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_RECONCILE *r, WT_M
...
-        /* Flush the updates on stack. */
-        for (; modifies.size > 0; tmp = full_value, full_value = prev_full_value,
-                                  prev_full_value = tmp, upd = prev_upd) {
+        /*
+         * Flush the updates on stack. Stopping once we run out or we reach the onpage upd start
+         * time pair, we can squash modifies with the same start time pair as the onpage upd away.
+         */
+        for (; modifies.size > 0 &&
+             !(upd->txnid == list->onpage_upd->txnid &&
+                 upd->start_ts == list->onpage_upd->start_ts);
```
Comment (quchenhao) [inline src/history/hs.c]: may be (upd->txnid != list->onpage_upd->txnid ||
                 upd->start_ts != list->onpage_upd->start_ts)

File: src/history/hs.c [hunk 4]
```diff
@@ -621,11 +582,10 @@ __wt_hs_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_RECONCILE *r, WT_M
             /*
              * Skip the updates have the same start timestamp and transaction id
              *
-             * The update older than onpage_upd can be squashed away. Insert a full update anyway to
-             * simplify the code. It will take some extra space but such case should be rare.
+             * Modifies that have the same start time pair as the onpage_upd can be squashed away.
              */
             if (upd->start_ts != prev_upd->start_ts || upd->txnid != prev_upd->txnid ||
-              modifies.size == 0) {
+              upd->start_ts != prev_upd->start_ts) {
```
Comment (quchenhao) [inline src/history/hs.c]: Redundant

**Comment by luke-pearson** [comment]
evergreen retry

**Pull request merged**